tests/e2e/objects/views/View.ts

```diff
@@ -0,0 +1,164 @@
+
+    if (options?.waitForItems) {
+      await this.waitForFilteredItems(locator, options.waitForItems, options);
+    }
+
+    const items: ViewItem[] = [];
+    const count: number = await locator.count();
+    for (let i = 0; i < count; i++) {
+      const element: Locator = locator.nth(i);
+      const item = new ViewItem(this.page, element);
+      items.push(item);
+    }
+    return items;
+  }
+
+  /** Build a locator with filters applied based on the provided options. */
+  private buildFilteredLocator(options?: {
+    text?: string | RegExp;
+    iconId?: string;
+    level?: number;
+  }): Locator {
+    let locator = this.body.locator('[role="treeitem"]');
+    if (options?.level) {
+      locator = this.body.locator(`[role="treeitem"][aria-level="${options.level}"]`);
+    }
+
+    if (options?.text) {
+      locator = locator.filter({ hasText: options.text });
+    }
+    if (options?.iconId) {
+      // we're already filtering the view body for tree items (maybe with a level), so
+      // we just need to look for the ones that have the specified icon
+      locator = locator.filter({ has: this.page.locator(`.codicon-${options.iconId}`) });
+    }
+    return locator;
+  }
+
+  /** Wait for filtered items to appear, with incremental checks for easier debugging. */
+  private async waitForFilteredItems(
+    locator: Locator,
+    waitConfig: { timeout?: number; minCount?: number } | boolean,
+    options?: {
+      text?: string | RegExp;
+      iconId?: string;
+      level?: number;
+    },
+  ): Promise<void> {
+    const config =
+      waitConfig === true ? {} : (waitConfig as { timeout?: number; minCount?: number });
+    const timeout = config.timeout ?? 10_000;
+    const minCount = config.minCount ?? 1;
+
+    // 1. wait for the view itself to be ready
+    await this.body.waitFor({ state: "visible", timeout });
+
+    // 2. wait for any tree item to be visible
+    const anyItemLocator = this.body.locator('[role="treeitem"]');
+    try {
+      await anyItemLocator.first().waitFor({ state: "visible", timeout });
+    } catch (e) {
+      throw new Error("No tree items found in view", { cause: e });
+    }
+
+    // 3. wait for the filtered locator to find a match
+    try {
+      await locator.first().waitFor({ state: "visible", timeout });
+    } catch (e) {
+      const allItemsCount = await anyItemLocator.count();
+      const allItemsText = await anyItemLocator.allInnerTexts();
+      throw new Error(
+        `No items found matching filter: ${JSON.stringify(
+          options,
+        )}. Found ${allItemsCount} total items: [${allItemsText.join(", ")}]`,
+        { cause: e },
+      );
+    }
+
+    const currentCount = await locator.count();
+    if (currentCount < minCount) {
+      const allItemsText = await locator.allInnerTexts();
+      throw new Error(
+        `Expected at least ${minCount} items, but found ${currentCount}: [${allItemsText.join(
+          ", ",
+        )}]`,
+      );
+    }
```

Comment:
Warning, `count()` call is not auto awaitable, which means flakiness, at least theoretically.